math-scala/src/main/scala/org/apache/mahout/math/indexeddataset/BiMap.scala

```diff
@@ -0,0 +1,122 @@
+ * @param m Map to use for forward reference
+ * @param i optional reverse map of value to key, will create one in the constructor if none is provided
+ *          and is required to have no duplicate reverse mappings.
+ */
+class BiMap[K, V] (
+    private val m: Map[K, V],
+    private val i: Option[BiMap[V, K]] = None
+  ) extends Serializable {
+
+  // NOTE: make inverse's inverse point back to current BiDictionary
+  val inverse: BiMap[V, K] = i.getOrElse {
+    val rev = m.map(_.swap)
+    require((rev.size == m.size),
+      s"Failed to create reversed map. Cannot have duplicated values.")
+    new BiMap(rev, Some(this))
+  }
+
+  def get(k: K): Option[V] = m.get(k)
+
+  def getOrElse(k: K, default: => V): V = m.getOrElse(k, default)
+
+  def contains(k: K): Boolean = m.contains(k)
+
+  def apply(k: K): V = m.apply(k)
+
+  /**
+   * Converts to a map.
+   * @return a map of type immutable.Map[K, V]
+   */
+  def toMap: Map[K, V] = m
+
+  /**
+   * Converts to a sequence.
+   * @return a sequence containing all elements of this map
+   */
+  def toSeq: Seq[(K, V)] = m.toSeq
+
+  def size: Int = m.size
+
+  def take(n: Int) = BiMap(m.take(n))
+
+  override def toString = m.toString
+}
+
+object BiMap {
+
+  /** Extra constructor from a map */
+  def apply[K, V](x: Map[K, V]): BiMap[K, V] = new BiMap(x)
+
+}
+
+/** BiDictionary is a specialized BiMap that has non-negative Ints as values for use as DRM keys */
+class BiDictionary (
+    private val m: Map[String, Int],
+    private val i: Option[BiMap[Int, String]] = None )
+  extends BiMap[String, Int](m, i) {
+}
+
+/** BiDictionary is a specialized BiMap that has non-negative Ints as values for use as DRM keys.
+  * The companion object provides modification methods specific to maintaining contiguous Int values
+  * and unique String keys */
+object BiDictionary {
+
+  /**
+   * Append new keys to an existing BiDictionary and return the result. The values will start
+   * at m.size and increase to create a continuous non-zero value set from 0 to size - 1
+   * @param keys new keys to append
+   * @param biDi append keys to this BiDictionary and create new values buy incremeting from the highest Int value
+   * @return a BiDictionary with added mappings
+   */
+  def append(keys: Set[String], biDi: BiDictionary): BiDictionary = {
+    val hm = HashMap(keys.toSeq.view.zip (Stream from biDi.size): _*)
+    new BiDictionary(biDi.m ++ hm)
+  }
+
+  /**
+   * Append new keys to an existing BiDictionary and return the result. The values will start
+   * at m.size and increase to create a continuous non-zero value set from 0 to size - 1
+   * @param keys new keys to append
+   * @param biDi append keys to this BiDictionary and create new values buy incremeting from the highest Int value
+   * @return a BiDictionary with added mappings
+   */
+  def append(keys: List[String], biDi: BiDictionary): BiDictionary = {
+    val hm = HashMap(keys.view.zip (Stream from biDi.size): _*)
```

Comment:
Seems like a Seq will do for most collections of String so only one append method, I'll change and test. Not sure if there are significant performance issues. The use of .view is very subtle, not sure if it has any effect.